Improve compliance with current HTML and webappapis specifications for PDF Viewer

brentfulgham opened the pull request:
#### e6adcfb74700125e84be81d96f9ee16248145c82
<pre>
Improve compliance with current HTML and webappapis specifications for PDF Viewer
<a href="https://bugs.webkit.org/show_bug.cgi?id=245396">https://bugs.webkit.org/show_bug.cgi?id=245396</a>
&lt;rdar://100142265&gt;

Reviewed by Chris Dumez.

The HTML specification includes guidance for PDF Viewer handling (see
<a href="https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support).">https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support).</a>
We currently fail a number of tests in this area, even though we provide
proper features and handle PDF&apos;s properly.

We also failed to impleemnt the &apos;pdfViewerEnabled&apos; property on Navigator,
which might block sites in the future from offering inline PDF viewing.

This patch updates WebKit to supply the expected set of PDF Viewer plugins
(all aliases to the underlying PDF implementation) in the specified order
and with the specified descriptions and file names.

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Update for new passes.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/system-state-and-capabilities/the-navigator-object/plugins-and-mimetypes-expected.txt: Ditto.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Ditto.
* LayoutTests/platform/mac-wk1/fast/dom/navigator-detached-no-crash-expected.txt: Ditto.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Ditto.
* LayoutTests/platform/mac/fast/dom/navigator-detached-no-crash-expected.txt: Ditto.
* Source/WebCore/page/Navigator.cpp:
(WebCore::dummyPDFPluginNamess): Added.
(WebCore::Navigator::initializePluginAndMimeTypeArrays): Revise to generate
relevant dummy entries as defined by the standard.
(WebCore::Navigator::pdfViewerEnabled): Added.
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/NavigatorPlugins.idl: Add the new &apos;pdfViewerEnabled&apos; property.
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::pdfDocumentTypeDescription): Added implementation for non-Cocoa ports (which
use the pdf.js implementation).
* Source/WebCore/platform/LocalizedStrings.h
* Source/WebCore/plugins/PluginData.cpp:
(WebCore::PluginData::initPlugins): Capture the PDF viewer plugin (if present).
* Source/WebCore/plugins/PluginData.h:
(WebCore::PluginData::builtInPDFPlugIn const): Added.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::pluginInfo): Update to provide the correct description
and &apos;file&apos; label, as defined by the spec.

Canonical link: <a href="https://commits.webkit.org/255052@main">https://commits.webkit.org/255052@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b286364d5d5afb7a80730968ce035783f3ab9760

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91137 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/137 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/100646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/160109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/149 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96793 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/83499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/35285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/33081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36864 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1550 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38791 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,21 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
 
+constexpr size_t dummyPDFPluginCount = 4;
+
+static const Vector<String> dummyPDFPluginNamess(bool hasWebKitBuiltinPDFPlugIn)
```
typo: dummyPDFPluginNamess

cdumez replied:
No reason to return a *const* Vector. We can drop the `const`.

Unless you make the Vector static and then we should return a const reference to the vector.

brentfulgham replied:
Fixed!

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,21 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+{
+    Vector<String> dummyPluginNames;
+    
+    dummyPluginNames.append(makeString("PDF Viewer"));
```
Please never use makeString() to convert a string literal to a String. You're supposed to use the `_s` suffix, that's it.
So `makeString("PDF Viewer")` -> `"PDF Viewer"_s`.

Also, we can write this a bit more efficiently:
```
Vector<String> dummyPluginNames {
    "PDF Viewer"_s,
    "Chrome PDF Viewer"_s,
    "Microsoft Edge PDF Viewer"_s,
    "Microsoft Edge PDF Viewer"_s,
    "WebKit built-in PDF"_s,
};
```

We may even want to make it static since it doesn't change. If you make it static then the function should return a `const Vector<String>&`

brentfulgham replied:
Fixed!

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,21 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+    dummyPluginNames.append(makeString("Chrome PDF Viewer"));
+    dummyPluginNames.append(makeString("Chromium PDF Viewer"));
+    dummyPluginNames.append(makeString("Microsoft Edge PDF Viewer"));
+    if (!hasWebKitBuiltinPDFPlugIn)
```
The spec doesn't have this condition, it is either:
- An empty vector
or
- A vector containing all 5 strings in a specific order.

brentfulgham replied:
In the case where we are using the true WebKit Built-in Plugin, there is a real plugin with this name. We do not need the dummy for that case. But I can filter that out at a later step.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -247,11 +263,29 @@ void Navigator::initializePluginAndMimeTypeArrays()
     Vector<Ref<DOMMimeType>> webVisibleDOMMimeTypes;
 
-    publiclyVisibleDOMPlugins.reserveInitialCapacity(publiclyVisiblePlugins.size());
+    publiclyVisibleDOMPlugins.reserveInitialCapacity(publiclyVisiblePlugins.size() + dummyPDFPluginCount + (reportDummyPDFPlugin ? 1 : 0));
```
If `!m_pdfViewerEnabled` then we should not report the dummy PDF plugin names so this doesn't seem quite right.

brentfulgham replied:
Good point. Fixed.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -247,11 +263,29 @@ void Navigator::initializePluginAndMimeTypeArrays()
+    bool hasWebKitBuiltinPDFPlugin = frame->page()->pluginData().builtInPDFPlugin().has_value();
+    bool reportDummyPDFPlugin = frame->loader().client().canShowMIMEType("application/pdf"_s);
+
+    m_pdfViewerEnabled = hasWebKitBuiltinPDFPlugin || reportDummyPDFPlugin;
```
Is `hasWebKitBuiltinPDFPlugin` really needed? Wouldn't `frame->loader().client().canShowMIMEType("application/pdf"_s)` be true also if with had the builtin PDF plugin?

brentfulgham replied:
Yes. We are trying to distinguish between the case where we have a real plugin (macOS), and all the other platforms that either use a separate PDF view (iOS, WebKitLegacy), or PDFJS. So this flag is needed so we can populate

cdumez commented on Source/WebCore/plugins/PluginData.h:
```diff
@@ -113,6 +113,10 @@ class PluginData : public RefCounted<PluginData> {
 
     String pluginFileForWebVisibleMimeType(const String& mimeType) const;
 
+    std::optional<PluginInfo> builtInPDFPlugin() const { return m_builtInPDFPluginInfo; }
```
Should return a `const std::optional<PluginInfo>&` to avoid copying whenever possible.

brentfulgham replied:
Fixed.

cdumez commented on Source/WebCore/plugins/PluginData.cpp:
```diff
@@ -32,6 +32,8 @@
 
 namespace WebCore {
 
+static inline bool isBuiltInPDFPlugIn(const PluginInfo&);
```
Where is this implemented?

brentfulgham replied:
Later in the file. I'll move to this location, since it's a small function.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,25 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+// https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support
+// Section 8.9.1.6 states that if pdfViewerEnabled is true, we must return a list
+// of exactly five PDF view plugins, in a particular order.
+constexpr size_t dummyPDFPluginCount = 5;
```
Any reason we can't call `dummyPDFPluginNames().size()`?

brentfulgham replied:
No -- I'll change to that.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,25 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+static const Vector<String>& dummyPDFPluginNames()
+{
+    static NeverDestroyed<Vector<String>> dummyPluginNames(std::initializer_list<String> {
+        String { "PDF Viewer"_s },
```
Do we really need the `String { }`?

brentfulgham replied:
I thought I did (I copied this format from elsewhere), but I'll try without.

brentfulgham replied:
It seems to work properly without the String, so I will fix that.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,25 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+        String { "Chrome PDF Viewer"_s },
+        String { "Chromium PDF Viewer"_s },
+        String { "Microsoft Edge PDF Viewer"_s },
+        builtInPDFPluginName(),
```
I don't think we should be using `builtInPDFPluginName()` since this is a translated string. The spec says to use "WebKit built-in PDF". This would cause us to no longer be compliant on non-US systems.

brentfulgham replied:
Okay, I'll change that. Then I can remove the LocalizedString changes I made, too.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -247,11 +267,44 @@ void Navigator::initializePluginAndMimeTypeArrays()
+        // https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support
+        // Section 8.9.1.6 states that if pdfViewerEnabled is true, we must return a list
+        // of exactly five PDF view plugins, in a particular order.
+        const auto& builtinName = builtInPDFPluginName();
```
builtinName is a super generic name considering we're talking about the built-in PDF plugin specifically.

brentfulgham replied:
I'll revise that to use an `ASCIILiteral` to represent the 'WebKit built-in PDF' label, and use that in the various places that need to know the string.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -247,11 +267,44 @@ void Navigator::initializePluginAndMimeTypeArrays()
+
+    publiclyVisibleDOMPlugins.reserveInitialCapacity(publiclyVisibleDOMPluginCount);
+
+#if PLATFORM(COCOA) || ENABLE(PDFJS)
```
There is something odd here. You set `m_pdfViewerEnabled` to true potentially earlier.
The spec says that if m_pdfViewerEnabled is true, then the plugin list must contain the 5 dummy plugin names.
However, adding the dummy plugin names is behind a `#if PLATFORM(COCOA) || ENABLE(PDFJS)` check.

Therefore, non-Cocoa ports that return true for frame->loader().client().canShowMIMEType("application/pdf"_s) will not be standard-compliant since they will return true for `navigator.pdfViewerEnabled` but not have the dummy plugins names.

brentfulgham replied:
I'm not sure if we have a port that is non-Cocoa and also non-PDF.JS, but if we did it would not behave properly with my logic. I'll remove the compile guards.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,26 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+// https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support
+// Section 8.9.1.6 states that if pdfViewerEnabled is true, we must return a list
+// of exactly five PDF view plugins, in a particular order.
+constexpr size_t dummyPDFPluginCount = 5;
```
Unused, should be removed.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,26 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+// of exactly five PDF view plugins, in a particular order.
+constexpr size_t dummyPDFPluginCount = 5;
+
+const ASCIILiteral webKitBuiltInPDFPluginName { "WebKit built-in PDF"_s };
```
constexpr?

brentfulgham replied:
Done (but now removed).

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,26 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+constexpr size_t dummyPDFPluginCount = 5;
+
+const ASCIILiteral webKitBuiltInPDFPluginName { "WebKit built-in PDF"_s };
+const ASCIILiteral genericPDFViewerName { "PDF Viewer"_s };
```
ditto.

brentfulgham replied:
Done (and kept!)

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -251,7 +272,45 @@ void Navigator::initializePluginAndMimeTypeArrays()
+            --publiclyVisibleDOMPluginCount; // Don't double-count the PDF Plugin
+    }
+
+    publiclyVisibleDOMPlugins.reserveInitialCapacity(publiclyVisibleDOMPluginCount);
```
I think the logic is complex enough now that we want to stop using reserveInitialCapacity() + uncheckedAppend(), and just use append() instead. This is a security bug waiting to happen.

brentfulgham replied:
Done -- and simplified in a follow-up.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -265,11 +324,14 @@ void Navigator::initializePluginAndMimeTypeArrays()
-    std::sort(publiclyVisibleDOMPlugins.begin(), publiclyVisibleDOMPlugins.end(), [](const Ref<DOMPlugin>& a, const Ref<DOMPlugin>& b) {
+    ASSERT(!m_pdfViewerEnabled || publiclyVisibleDOMPlugins.begin()->get().info().name == genericPDFViewerName);
+    size_t dummyPDFOffset = m_pdfViewerEnabled ? 1 : 0;
+    std::sort(publiclyVisibleDOMPlugins.begin() + dummyPDFOffset, publiclyVisibleDOMPlugins.end(), [](const Ref<DOMPlugin>& a, const Ref<DOMPlugin>& b) {
```
Why this sorting? And why do we have additionalWebVisiblePlugins?

Per the spec, the list should either be empty (when pdf is not supported) or a dummy static list of 5 PDF plugins (when pdf is supported).

It doesn't look like we are really matching the spec here? Our logic looks terribly complicated for something that looks so simple in the spec.

brentfulgham replied:
All of that was a holdover from when we had many plugins, and wanted to always report the plugins in the same order regardless of when the user installed a plugin, or enabled/disabled them. This is no longer needed so I removed it in a follow-up.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -233,6 +234,24 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+// https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support
+// Section 8.9.1.6 states that if pdfViewerEnabled is true, we must return a list
+// of exactly five PDF view plugins, in a particular order.
+constexpr ASCIILiteral webKitBuiltInPDFPluginName { "WebKit built-in PDF"_s };
```
It looks like this is used only in one place so we may not need a global variable for this.

brentfulgham replied:
Good catch. I've removed it.

cdumez reviewed (approved):
LGTM assuming the bots are happy.

brentfulgham committed:
Improve compliance with current HTML and webappapis specifications for PDF Viewer
https://bugs.webkit.org/show_bug.cgi?id=245396
<rdar://100142265>

Reviewed by Chris Dumez.

The HTML specification includes guidance for PDF Viewer handling (see
https://html.spec.whatwg.org/multipage/system-state.html#pdf-viewing-support).
We currently fail a number of tests in this area, even though we provide
proper features and handle PDF's properly.

We also failed to impleemnt the 'pdfViewerEnabled' property on Navigator,
which might block sites in the future from offering inline PDF viewing.

This patch updates WebKit to supply the expected set of PDF Viewer plugins
(all aliases to the underlying PDF implementation) in the specified order
and with the specified descriptions and file names.

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Update for new passes.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/system-state-and-capabilities/the-navigator-object/plugins-and-mimetypes-expected.txt: Ditto.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Ditto.
* LayoutTests/platform/mac-wk1/fast/dom/navigator-detached-no-crash-expected.txt: Ditto.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt: Ditto.
* LayoutTests/platform/mac/fast/dom/navigator-detached-no-crash-expected.txt: Ditto.
* Source/WebCore/page/Navigator.cpp:
(WebCore::dummyPDFPluginNamess): Added.
(WebCore::Navigator::initializePluginAndMimeTypeArrays): Revise to generate
relevant dummy entries as defined by the standard.
(WebCore::Navigator::pdfViewerEnabled): Added.
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/NavigatorPlugins.idl: Add the new 'pdfViewerEnabled' property.
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::pdfDocumentTypeDescription): Added implementation for non-Cocoa ports (which
use the pdf.js implementation).
* Source/WebCore/platform/LocalizedStrings.h
* Source/WebCore/plugins/PluginData.cpp:
(WebCore::PluginData::initPlugins): Capture the PDF viewer plugin (if present).
* Source/WebCore/plugins/PluginData.h:
(WebCore::PluginData::builtInPDFPlugIn const): Added.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::pluginInfo): Update to provide the correct description
and 'file' label, as defined by the spec.

Canonical link: https://commits.webkit.org/255052@main
